btree/btree.c

```diff
@@ -5360,29 +5366,38 @@ static int64_t btree_del_kv_tick(struct m0_sm_op *smop)
 			}
 			return btree_del_resolve_underflow(bop);
 		}
-		lock_op_unlock(tree);
-		return m0_sm_op_sub(&bop->bo_op, P_CLEANUP, P_FINI);
+		M0_ASSERT(0);
 	}
+	case P_CAPTURE:
+		/**
+		 * TBD: uncomment function call below once node_free changes
+		 * done.
+		 */
+		/* btree_tx_node_capture(oi, bop->bo_tx); */
+		return P_FREENODE;
 	case P_FREENODE : {
-		struct nd *node;
-
-		lev = &oi->i_level[oi->i_used];
-		if (lev->l_freenode) {
-			M0_ASSERT(oi->i_used > 0);
-			oi->i_used --;
-			node = lev->l_node;
-			lev->l_node = NULL;
+		int i;
+		for (i = oi->i_used; i >= 0; i--) {
+			lev = &oi->i_level[i];
+			if (lev->l_freenode) {
+				M0_ASSERT(oi->i_used > 0);
+				oi->i_nop.no_opc = NOP_FREE;
+				node_free(&oi->i_nop, lev->l_node,
```

Comment:
Where is the call to node_fini() as we discussed. It needs to be done before we reach this point.

Reply:
I will be covering it different PR as it may require some more changes.